src/Compilers/Core/Rebuild/BuildConstructor.cs

```diff
@@ -30,74 +30,64 @@ public class BuildConstructor
     {
         private readonly ILogger _logger;
 
+        // TODO: shouldn't need to pass a logger.
         public BuildConstructor(ILogger logger)
         {
             _logger = logger;
         }
 
-        public (Compilation? compilation, bool isError) CreateCompilation(
+        public Compilation CreateCompilation(
+            string assemblyFileName,
             CompilationOptionsReader compilationOptionsReader,
-            string fileName,
-            ImmutableArray<ResolvedSource> sources,
+            ImmutableArray<SyntaxTreeInfo> syntaxTreeInfos,
             ImmutableArray<MetadataReference> metadataReferences)
         {
-            // We try to handle assemblies missing compilation options gracefully by skipping them.
-            // However, if an assembly has some bad combination of data, for example if it contains
-            // compilation options but not metadata references, then we throw an exception.
-            if (!compilationOptionsReader.TryGetMetadataCompilationOptions(out var pdbCompilationOptions)
-                || pdbCompilationOptions.Length == 0)
+            var diagnosticBag = DiagnosticBag.GetInstance();
+            var compilation = compilationOptionsReader.GetLanguageName() switch
             {
-                _logger.LogInformation($"{fileName} did not contain compilation options in its PDB");
-                return (compilation: null, isError: false);
-            }
+                LanguageNames.CSharp => CreateCSharpCompilation(assemblyFileName, compilationOptionsReader, syntaxTreeInfos, metadataReferences),
+                LanguageNames.VisualBasic => CreateVisualBasicCompilation(assemblyFileName, compilationOptionsReader, syntaxTreeInfos, metadataReferences, diagnosticBag),
+                _ => throw new InvalidDataException($"{assemblyFileName} is not a known language")
```

Comment:
>$"{assemblyFileName} is not a known language" [](start = 52, length = 45)

This message may be confusing. Consider including the language: $"{assemblyFileName} unsupported language: {language}".